Add fmtlib dependency

feltech opened the pull request:
## Description

Closes #1070.

fmtlib allows for easy, safe string substitution in C++, adding advanced formatting functionality (e.g. rounding), and reducing bloat.

So add as a new private build-only header-only dependency.

Using header-only increases compile times, but subjectively it's not noticeable, and has the benefit of avoiding a possible public library dependency (if OpenAssetIO is built as a static library).

Tweak a single existing, tested, log string to make use of libfmt, to ensure basic usage is validated.

Added to openassetio-build Docker image: https://github.com/OpenAssetIO/OpenAssetIO/pkgs/container/openassetio-build

- [x] I have updated the release notes.
- ~~[ ] I have updated all relevant user documentation.~~

## Reviewer Notes

<!--- Provide any notes to the reviewer that might help them more easily
      understand the changeset. --->

## Test Instructions

<!--- Provide instructions to the reviewer on how to explicitly test
      these changes. --->

foundrytom commented:
I know we removed `CONDING_STANDARDS` as there is only one way to use the header version - but its it worth adding a more general note that "we use `libfmt` for string building wherever needed, do not use `stringstream` etc..."

feltech commented on src/openassetio-core/src/hostApi/Manager.cpp:
```diff
@@ -53,12 +55,11 @@ std::optional<Str> entityReferencePrefixFromInfo(const log::LoggerInterfacePtr &
-          "' provided by manager's info() dict. Subsequent calls to isEntityReferenceString will"
-          " use this prefix rather than call the manager's implementation.";
-      logger->debugApi(msg);
+      logger->debugApi(
+          fmt::format("Entity reference prefix '{}' provided by manager's info() dict. Subsequent"
+                      " calls to isEntityReferenceString will use this prefix rather than call the"
+                      " manager's implementation.",
+                      *prefixPtr));
```
That's an interesting idea, so logger methods become a wrapper around `fmt::format`, and we can skip the interpolation if the log severity is too low. :thinking: tricky actually, since we can't use variadic templates for a `virtual` method.  Worth looking into in case there's a way to do it.

feltech replied:
A macro, or even a normal function, that wraps the `LoggerInterface` as a middleware, could do it I think, yeah.

foundrytom replied:
Ace!

feltech commented:
Added an emoji to the release notes to signal a build dependency, see what you think (open to ideas on alternative emojis).

> I know we removed `CONDING_STANDARDS` as there is only one way to use the header version - but its it worth adding a more general note that "we use `libfmt` for string building wherever needed, do not use `stringstream` etc..."

Added (force-pushed).

feltech commented:
Fixup pushed to reduce fmtlib version to 9.1.0, which doesn't export errant symbols.  Also included the `Makefile` change for the Docker container.

foundrytom commented:
> Fixup pushed to reduce fmtlib version to 9.1.0, which doesn't export errant symbols. Also included the `Makefile` change for the Docker container.

Thanks for posting upstream does [this comment](https://github.com/fmtlib/fmt/issues/3626#issuecomment-1707240105) imply we may have issues on Windows with the older version?

feltech commented:
> > Fixup pushed to reduce fmtlib version to 9.1.0, which doesn't export errant symbols. Also included the `Makefile` change for the Docker container.
> 
> Thanks for posting upstream does [this comment](https://github.com/fmtlib/fmt/issues/3626#issuecomment-1707240105) imply we may have issues on Windows with the older version?

As discussed offline - posted a follow-up question upstream to make sure I understand, but pretty certain we don't have a problem, especially in header-only mode.

foundrytom commented on resources/build/Makefile:
```diff
@@ -11,7 +11,7 @@ default: docker-image
 
 # This container version mimicks the VFX reference platform versioning.
-CONTAINER_VERSION=2022.1
+CONTAINER_VERSION=2022.2
```
[fix] As this has been published, should this be .3? (and do we need a commit somewhere with as it was for the published .2 container)

feltech replied:
It was published as .1, I bumped to .2.  I overwrote the previous published .2 with the downgraded fmtlib, reasoning that it's unlikely anyone pulled .2 in those couple of hours.

Or do you mean we should bump to the not-yet-released future version inside the Makefile?

foundrytom replied:
If you've already overwritten the previous publish, then its moot, (good to avoid that where we can next time).

feltech commented on doc/contributing/CODING_STANDARDS.md:
```diff
@@ -176,6 +176,13 @@ when binding a C++ `enum class` using pybind, use
+The project has a (private, header-only) dependency on the
+[fmt](https://fmt.dev/9.1.0) library for efficient string formatting,
+and this should be used by preference over alternative legacy options
+such as `std::stringstream`.
```
Added in f7441be5

foundrytom reviewed (approved):
Assuming were happy with the symbol situation, then LGTM 👍 🙏
